Hint about missing trait method suggests incorrect code (self parameter not named "self").

In the error message below, the first argument of  `poll` is not named `self`.

```rust
use core::future::Future;

struct S;
impl Future for S {}
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8760b4f73c3bc7e383edf62742ccbac5))

Errors:

```
   Compiling playground v0.0.1 (/playground)
error[E0046]: not all trait items implemented, missing: `Output`, `poll`
 --> src/lib.rs:4:1
  |
4 | impl Future for S {}
  | ^^^^^^^^^^^^^^^^^ missing `Output`, `poll` in implementation
  |
  = help: implement the missing item: `type Output = Type;`
  = help: implement the missing item: `fn poll(_: std::pin::Pin<&mut Self>, _: &mut std::task::Context<'_>) -> std::task::Poll<<Self as std::future::Future>::Output> { unimplemented!() }`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0046`.
error: could not compile `playground`.

To learn more, run the command again with --verbose.

```

@rustbot modify labels: T-compiler, A-diagnostics, C-enhancement, D-papercut, F-arbitrary_self_types.

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->
This issue has been assigned to @Duddino via [this comment](https://github.com/rust-lang/rust/issues/71150#issuecomment-614260066).
<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"Duddino"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

I’m not just interested in a fix for this particular case. Types that allow dispatch can be arbitrarily nested AFAIK and also I didn’t tag `F-arbitrary_self_types` for no reason - these can be extended. Furthermore the correct way to fix this is probably _not_ to just add code but rather to change ~existing logic that must be somewhere already, for example some code currently _does_ make `self`, `&self` and `&mut self` parameters print correctly and differently from non-self parameters that have type `Self` or~...... _let me double-check that...._

_Oh, wow, I’ll expand this bug-report._

## Hint about missing trait method suggests incorrect code (_non_-self parameter _mistakenly_ named "self").

Create a binary crate named “my-crate” with:

**main.rs**
```rust
struct S;
impl my_crate::Tr for S {}

fn main() {
    println!("Hello, world!");
}
```

**lib.rs**
```rust
pub trait Tr {
  fn f(self: &Self, not_self: &Self);
}
```

Error message is:
```
error[E0046]: not all trait items implemented, missing: `f`
 --> src\main.rs:2:1
  |
2 | impl my_crate::Tr for S {}
  | ^^^^^^^^^^^^^^^^^^^^^^^ missing `f` in implementation
  |
  = help: implement the missing item: `fn f(&self, &self) { unimplemented!() }`
```

--- a/src/librustc_typeck/check/mod.rs
+++ b/src/librustc_typeck/check/mod.rs
@@ -2267,7 +2267,7 @@ fn fn_sig_suggestion(
                     };
                     if assoc.fn_has_self_parameter && i == 0 {
                         format!("&{}{}self", reg, mutability.prefix_str())
-                    }else {
+                    } else {
                         format!("_: {:?}", ty)
                     }
                 }

--- a/src/librustc_typeck/check/mod.rs
+++ b/src/librustc_typeck/check/mod.rs
@@ -2251,26 +2251,33 @@ fn fn_sig_suggestion(
     sig: &ty::FnSig<'_>,
     ident: Ident,
     predicates: ty::GenericPredicates<'_>,
+    assoc: &ty::AssocItem,
 ) -> String {
     let args = sig
         .inputs()
         .iter()
-        .map(|ty| {
+        .enumerate()
+        .map(|(i, ty)| {
             Some(match ty.kind {
-                ty::Param(param) if param.name == kw::SelfUpper => "self".to_string(),
-                ty::Ref(reg, ref_ty, mutability) => {
+                ty::Param(_) if assoc.fn_has_self_parameter && i == 0 => "self".to_string(),
+                ty::Ref(reg, _ref_ty, mutability) => {
                     let reg = match &format!("{}", reg)[..] {
                         "'_" | "" => String::new(),
                         reg => format!("{} ", reg),
                     };
-                    match ref_ty.kind {
-                        ty::Param(param) if param.name == kw::SelfUpper => {
-                            format!("&{}{}self", reg, mutability.prefix_str())
-                        }
-                        _ => format!("_: {:?}", ty),
+                    if assoc.fn_has_self_parameter && i == 0 {
+                        format!("&{}{}self", reg, mutability.prefix_str())
+                    }else {
+                        format!("_: {:?}", ty)
                     }
                 }
-                _ => format!("_: {:?}", ty),
+                _ => {
+                    if assoc.fn_has_self_parameter && i == 0 {
+                        format!("self: {:?}", ty)
+                    } else {
+                        format!("_: {:?}", ty)
+                    }
+                }
             })
         })
         .chain(std::iter::once(if sig.c_variadic { Some("...".to_string()) } else { None }))
@@ -2309,6 +2316,7 @@ fn suggestion_signature(assoc: &ty::AssocItem, tcx: TyCtxt<'_>) -> String {
                 tcx.fn_sig(assoc.def_id).skip_binder(),
                 assoc.ident,
                 tcx.predicates_of(assoc.def_id),
+                assoc,
             )
         }
         ty::AssocKind::Type => format!("type {} = Type;", assoc.ident),

--- a/src/librustc_typeck/check/mod.rs
+++ b/src/librustc_typeck/check/mod.rs
@@ -2260,13 +2260,19 @@ fn fn_sig_suggestion(
         .map(|(i, ty)| {
             Some(match ty.kind {
                 ty::Param(_) if assoc.fn_has_self_parameter && i == 0 => "self".to_string(),
-                ty::Ref(reg, _ref_ty, mutability) => {
+                ty::Ref(reg, ref_ty, mutability) if i == 0 => {
                     let reg = match &format!("{}", reg)[..] {
                         "'_" | "" => String::new(),
                         reg => format!("{} ", reg),
                     };
-                    if assoc.fn_has_self_parameter && i == 0 {
-                        format!("&{}{}self", reg, mutability.prefix_str())
+                    if assoc.fn_has_self_parameter {
+                        match ref_ty.kind {
+                            ty::Param(param) if param.name == kw::SelfUpper => {
+                                format!("&{}{}self", reg, mutability.prefix_str())
+                            }
+
+                            _ => format!("self: {}", ty),
+                        }
                     } else {
                         format!("_: {:?}", ty)
                     }

--- a/src/test/ui/missing/missing-items/auxiliary/m1.rs
+++ b/src/test/ui/missing/missing-items/auxiliary/m1.rs
@@ -2,4 +2,7 @@ pub trait X {
     const CONSTANT: u32;
     type Type;
     fn method(&self, s: String) -> Self::Type;
+    fn method2(self: Box<Self>, s: String) -> Self::Type;
+    fn method3(other: &Self, s: String) -> Self::Type;
+    fn method4(&self, other: &Self) -> Self::Type;
 }

--- a/src/test/ui/missing/missing-items/auxiliary/m1.rs
+++ b/src/test/ui/missing/missing-items/auxiliary/m1.rs
@@ -5,4 +5,5 @@ pub trait X {
     fn method2(self: Box<Self>, s: String) -> Self::Type;
     fn method3(other: &Self, s: String) -> Self::Type;
     fn method4(&self, other: &Self) -> Self::Type;
+    fn method5(self: &Box<Self>) -> Self::Type;
 }

--- a/src/test/ui/missing/missing-items/m2.stderr
+++ b/src/test/ui/missing/missing-items/m2.stderr
@@ -1,8 +1,8 @@
-error[E0046]: not all trait items implemented, missing: `CONSTANT`, `Type`, `method`, `method2`, `method3`, `method4`
+error[E0046]: not all trait items implemented, missing: `CONSTANT`, `Type`, `method`, `method2`, `method3`, `method4`, `method5`
   --> $DIR/m2.rs:9:1
    |
 LL | impl m1::X for X {
-   | ^^^^^^^^^^^^^^^^ missing `CONSTANT`, `Type`, `method`, `method2`, `method3`, `method4` in implementation
+   | ^^^^^^^^^^^^^^^^ missing `CONSTANT`, `Type`, `method`, `method2`, `method3`, `method4`, `method5` in implementation
    |
    = help: implement the missing item: `const CONSTANT: u32 = 42;`
    = help: implement the missing item: `type Type = Type;`
@@ -10,6 +10,7 @@ LL | impl m1::X for X {
    = help: implement the missing item: `fn method2(self: std::boxed::Box<Self>, _: std::string::String) -> <Self as m1::X>::Type { todo!() }`
    = help: implement the missing item: `fn method3(_: &Self, _: std::string::String) -> <Self as m1::X>::Type { todo!() }`
    = help: implement the missing item: `fn method4(&self, _: &Self) -> <Self as m1::X>::Type { todo!() }`
+   = help: implement the missing item: `fn method5(self: &std::boxed::Box<Self>) -> <Self as m1::X>::Type { todo!() }`
 
 error: aborting due to previous error
 

--- a/src/test/ui/missing/missing-items/m2.stderr
+++ b/src/test/ui/missing/missing-items/m2.stderr
@@ -1,12 +1,15 @@
-error[E0046]: not all trait items implemented, missing: `CONSTANT`, `Type`, `method`
+error[E0046]: not all trait items implemented, missing: `CONSTANT`, `Type`, `method`, `method2`, `method3`, `method4`
   --> $DIR/m2.rs:9:1
    |
 LL | impl m1::X for X {
-   | ^^^^^^^^^^^^^^^^ missing `CONSTANT`, `Type`, `method` in implementation
+   | ^^^^^^^^^^^^^^^^ missing `CONSTANT`, `Type`, `method`, `method2`, `method3`, `method4` in implementation
    |
    = help: implement the missing item: `const CONSTANT: u32 = 42;`
    = help: implement the missing item: `type Type = Type;`
    = help: implement the missing item: `fn method(&self, _: std::string::String) -> <Self as m1::X>::Type { todo!() }`
+   = help: implement the missing item: `fn method2(self: std::boxed::Box<Self>, _: std::string::String) -> <Self as m1::X>::Type { todo!() }`
+   = help: implement the missing item: `fn method3(_: &Self, _: std::string::String) -> <Self as m1::X>::Type { todo!() }`
+   = help: implement the missing item: `fn method4(&self, _: &Self) -> <Self as m1::X>::Type { todo!() }`
 
 error: aborting due to previous error
 

